SparseFixedBitSet#firstDoc: reduce number of `indices` iterations for a bit set that is not fully built yet.

epotyom opened the pull request:
In SparseFixedBitSet.firstDoc, instead of iterating though the entire indices array until non-zero value is found, keep track of max updated index.

Use case where it improves performance:
1. `SparseFixedBitSet` is created with high enough length, e.g. max doc in a segment
2. `#nextSetBit` is called (in a loop) on a bit set that is still being built, i.e. some of the next bits are `#set`, but the rest of the bit set is still empty.
3. The moment there are no further set bits, `#nextSetBit` call to `#firstDoc` iterates through the rest of `indices` array. 

In my case, we use SparseFixedBitSet to track and iterate children hits found in `ToParentBlockJoinQuery`. Iterating through empty `indices` elements becomes expensive when we do it for each parent docID.

Lucene util performance test results might not be great though - so maybe there is better way to achieve similar effect?

```
python3 src/python/localrun.py -source wikimediumall
...

                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
            BrowseDateTaxoFacets        1.65      (9.1%)        1.59      (0.5%)   -3.6% ( -12% -    6%) 0.081
       BrowseDayOfYearTaxoFacets        1.67      (9.2%)        1.61      (0.6%)   -3.5% ( -12% -    6%) 0.088
            MedTermDayTaxoFacets        9.40      (6.2%)        9.19      (5.0%)   -2.2% ( -12% -    9%) 0.206
     BrowseRandomLabelTaxoFacets        1.29      (4.6%)        1.27      (1.0%)   -1.9% (  -7% -    3%) 0.070
                         Prefix3      543.93      (5.7%)      535.20      (4.6%)   -1.6% ( -11% -    9%) 0.326
                      AndHighLow      780.30      (3.9%)      771.77      (4.1%)   -1.1% (  -8% -    7%) 0.383
                      AndHighMed      199.79      (2.3%)      197.77      (3.0%)   -1.0% (  -6% -    4%) 0.233
                 MedSloppyPhrase       61.79      (4.1%)       61.24      (4.1%)   -0.9% (  -8% -    7%) 0.488
                     AndHighHigh       84.66      (6.6%)       83.92      (7.6%)   -0.9% ( -14% -   14%) 0.699
                        PKLookup      143.72      (1.9%)      142.48      (2.1%)   -0.9% (  -4% -    3%) 0.171
                          Fuzzy1       56.85      (1.5%)       56.36      (2.0%)   -0.8% (  -4% -    2%) 0.122
            BrowseDateSSDVFacets        0.43     (16.7%)        0.43     (16.1%)   -0.8% ( -28% -   38%) 0.873
                        Wildcard      159.45      (2.7%)      158.30      (4.0%)   -0.7% (  -7% -    6%) 0.505
                          Fuzzy2       56.79      (1.2%)       56.38      (1.8%)   -0.7% (  -3% -    2%) 0.139
                      HighPhrase       20.07      (4.4%)       19.94      (5.8%)   -0.6% ( -10% -    9%) 0.701
                     MedSpanNear       15.66      (1.8%)       15.60      (2.2%)   -0.4% (  -4% -    3%) 0.537
                    OrNotHighMed      211.86      (3.1%)      211.03      (2.7%)   -0.4% (  -5% -    5%) 0.670
            HighTermTitleBDVSort       16.31      (2.8%)       16.25      (2.7%)   -0.4% (  -5% -    5%) 0.661
                       MedPhrase      154.39      (2.7%)      154.01      (3.3%)   -0.2% (  -6% -    5%) 0.800
                       OrHighMed      184.54      (2.5%)      184.21      (2.0%)   -0.2% (  -4% -    4%) 0.797
                       LowPhrase       72.18      (3.6%)       72.06      (4.3%)   -0.2% (  -7% -    8%) 0.893
                   OrHighNotHigh      229.39      (4.4%)      229.05      (4.5%)   -0.1% (  -8% -    9%) 0.915
                 LowSloppyPhrase       98.92      (1.5%)       98.84      (2.1%)   -0.1% (  -3% -    3%) 0.897
                     LowSpanNear       53.22      (1.0%)       53.21      (0.8%)   -0.0% (  -1% -    1%) 0.932
                         Respell       34.18      (1.9%)       34.18      (2.4%)   -0.0% (  -4% -    4%) 0.986
                    HighSpanNear        5.05      (3.1%)        5.06      (3.0%)    0.1% (  -5% -    6%) 0.929
         AndHighMedDayTaxoFacets       16.78      (1.6%)       16.79      (1.7%)    0.1% (  -3% -    3%) 0.850
                       OrHighLow      381.10      (3.4%)      381.52      (3.0%)    0.1% (  -6% -    6%) 0.914
                HighSloppyPhrase       12.20      (3.4%)       12.22      (4.1%)    0.1% (  -7% -    7%) 0.902
               HighTermMonthSort     1059.29      (4.7%)     1061.27      (4.8%)    0.2% (  -8% -   10%) 0.901
        AndHighHighDayTaxoFacets       13.53      (1.6%)       13.56      (1.7%)    0.2% (  -3% -    3%) 0.703
                    OrNotHighLow      664.93      (3.1%)      666.40      (3.6%)    0.2% (  -6% -    7%) 0.835
                         MedTerm      330.39      (8.8%)      331.13      (6.3%)    0.2% ( -13% -   16%) 0.927
                    OrHighNotLow      305.23      (5.3%)      306.22      (5.1%)    0.3% (  -9% -   11%) 0.844
     BrowseRandomLabelSSDVFacets        1.64      (4.1%)        1.65      (4.5%)    0.4% (  -7% -    9%) 0.754
                   OrNotHighHigh      170.52      (6.8%)      171.39      (6.2%)    0.5% ( -11% -   14%) 0.804
          OrHighMedDayTaxoFacets        3.03      (4.3%)        3.04      (3.1%)    0.6% (  -6% -    8%) 0.632
                         LowTerm      367.63      (5.1%)      370.21      (4.0%)    0.7% (  -7% -   10%) 0.629
                      TermDTSort       40.64      (4.8%)       41.01      (2.9%)    0.9% (  -6% -    9%) 0.467
                        HighTerm      321.26      (8.6%)      324.29      (6.8%)    0.9% ( -13% -   17%) 0.702
                      OrHighHigh      103.68      (8.6%)      104.72      (7.8%)    1.0% ( -14% -   19%) 0.699
             LowIntervalsOrdered       89.24      (4.8%)       90.40      (5.1%)    1.3% (  -8% -   11%) 0.409
             MedIntervalsOrdered       14.36      (5.8%)       14.59      (5.7%)    1.6% (  -9% -   13%) 0.386
            HighIntervalsOrdered       20.94      (4.8%)       21.28      (5.0%)    1.6% (  -7% -   11%) 0.299
               HighTermTitleSort       65.54      (2.8%)       66.62      (3.0%)    1.6% (  -4% -    7%) 0.073
                    OrHighNotMed      326.83      (3.5%)      332.80      (4.0%)    1.8% (  -5% -    9%) 0.127
           HighTermDayOfYearSort      313.09      (3.0%)      318.88      (4.2%)    1.8% (  -5% -    9%) 0.108
       BrowseDayOfYearSSDVFacets        2.58      (6.4%)        2.63      (6.3%)    1.9% ( -10% -   15%) 0.337
           BrowseMonthSSDVFacets        2.64      (6.8%)        2.71      (6.5%)    2.3% ( -10% -   16%) 0.268
           BrowseMonthTaxoFacets        1.81     (10.6%)        1.88     (10.3%)    3.6% ( -15% -   27%) 0.278
                          IntNRQ      134.47     (20.5%)      139.69     (16.0%)    3.9% ( -27% -   50%) 0.505
```

msokolov commented:
I wonder if `DocIdSetBuilder` would help?

msokolov commented:
Another idea -- would it help your use case? -- would be to support `nextSetBit(start, end)` . We could do this without adding any additional tracking in existing SparseBitSet methods

epotyom commented:
@msokolov thanks for looking into it!

> I wonder if DocIdSetBuilder would help? I guess it doesn't support accessing as a BitSet while under construction though.

Yes, I think this is exactly the reason we don't use it.

> Still maybe a similar class inspired from that, or even a fork of SparseFixedBitset could be helpful, although I do wonder how often SPFBS has a large number of trailing zeros.

Yeah I was curious if making in SparseFixedBitset makes sense, but I haven't found similar use cases in Lucene code. Also looking at benchmark results, it is probably a case for creating a fork.

> Another idea -- would it help your use case? -- would be to support nextSetBit(start, end) . We could do this without adding any additional tracking in existing SparseBitSet methods. Basically instead of asking SBS to keep track, the client might be able to do so?

Ah, I like this idea! Working on it. Not sure if it worth adding an abstract method to the base BitSet class, so I'll start with SparseFixedBitSet itself and see how it goes.

epotyom commented on lucene/core/src/java/org/apache/lucene/util/SparseFixedBitSet.java:
```diff
@@ -51,6 +51,7 @@ private static int blockCount(int length) {
+  int maxIndicesIndex = -1;
   final long[][] bits;
   final int length;
   int nonZeroLongCount;
```
Hmm, when we call `nextSetBit(int i)` I think we need to somehow know how many non-zero hits are *after* `i`, not just overall `nonZeroLongCount`; or maybe I'm missing something?

msokolov replied:
oh you're right! Somehow I assumed we would know how many non-zero longs were before `i` but now I see that's not the case.

gsmiller commented:
> Another idea -- would it help your use case? -- would be to support nextSetBit(start, end) . We could do this without adding any additional tracking in existing SparseBitSet methods. Basically instead of asking SBS to keep track, the client might be able to do so?


++ I like this suggestion as well but I wonder if the caller is always in a position to keep track of an `end`? It sounds like in your `TPBJQ` case though it's easy for the caller to know the range they're interested in, so maybe that's the right step here.

epotyom committed:
Add BitSet#firstSetBitInRange method and use it in BlockJoinSelector (#13559)

epotyom commented:
I've added new `BitSet#firstSetBitInRange` method with maybe naive implementations for both `SparseFixedBitSet` and `FixedBitSet`. I've also changed `BlockJoinSelector` to use it. `luceneutil` benchmark results seem to be neutral, but as far as I understand there are no tasks that trigger `BlockJoinSelector`?

I'll run our internal tests to see if it helps, and will add a line to CHANGES if otherwise the change looks good.

gsmiller reviewed (commented):
Overall looks good to me. Left a couple small comments. Only one that I have much of a preference on though, around whether-or-not the upper limit in the API should be inclusive vs. exclusive. Let's discuss that? Otherwise I like it. Thanks!

epotyom commented on lucene/core/src/java/org/apache/lucene/util/BitSet.java:
```diff
@@ -92,6 +92,12 @@ public void clear() {
   public abstract int nextSetBit(int index);
 
+  /**
+   * Returns the index of the first set bit from start (inclusive) until end (inclusive). {@link
```
Ah, makes sense, will change that.

gsmiller replied:
Also... looking at the `#substring` method signature got me thinking about the API name. We if we reuse the `nextSetBit` naming for the new method since it's essentially the same thing but with an early-termination optimization. We could just overload that method signature with `#nextSetBit(int beginIndex, int endIndex)`. We could also later overload `#prevSetBit` in the same way if there's ever a real use-case. Pretty minor suggestion but might be a nice naming convention. Thoughts?

epotyom replied:
I agree, overloading `nextSetBit` makes sense.

epotyom commented on lucene/join/src/java/org/apache/lucene/search/join/BlockJoinSelector.java:
```diff
@@ -64,14 +64,19 @@ public boolean get(int docID) {
+        final int lastPotentialChild = docID - 1;
+
+        final int firstPotentialChild = parents.prevSetBit(lastPotentialChild) + 1;
+        if (firstPotentialChild < docID) {
+          for (int child = children.firstSetBitInRange(firstPotentialChild, lastPotentialChild);
+              child != DocIdSetIterator.NO_MORE_DOCS;
+              child = children.firstSetBitInRange(child + 1, lastPotentialChild)) {
+            if (docsWithValue.get(child)) {
+              return true;
+            }
```
I'll apply the suggestion, thanks!

epotyom commented on lucene/core/src/java/org/apache/lucene/util/FixedBitSet.java:
```diff
@@ -291,6 +291,32 @@ public int nextSetBit(int index) {
+      return res > upperBound ? DocIdSetIterator.NO_MORE_DOCS : res;
+    }
+
+    int maxWord = Math.min((upperBound >> 6) + 1, numWords);
```
I've ended up removing `Math.min` here and making `upperBound <= bit set size` a requirement for the method - there is similar requirement for `index` parameter. Will push a commit shortly - please let me know what you think.

gsmiller replied:
Ah, that's even better. Thanks!

epotyom commented on lucene/core/src/java/org/apache/lucene/util/SparseFixedBitSet.java:
```diff
@@ -353,14 +355,47 @@ public int nextSetBit(int i) {
+    final long index = indices[i4096];
+    final long[] bitArray = this.bits[i4096];
+    int i64 = start >>> 6;
+    int o = Long.bitCount(index & ((1L << i64) - 1));
```
Make sense!

epotyom committed:
Address Greg's comments

epotyom committed:
Add CHANGES.txt entry

gsmiller reviewed (approved):
Looks good. Thanks @epotyom !

gsmiller committed:
Merge branch 'main' into SparseFixedBitSet-firstDoc

gsmiller commented:
@epotyom just merged. Thanks for the change! I just now noticed you put the changes entry under 10.0 but I don't see any reason we can't backport to 9.12. I'm going to backport now and move the changes entry but please let me know if there's some reason you think this should wait for 10. Thanks!
